### PR TITLE
Make `Updater.stop` Independent of `CancelledError`

### PR DESCRIPTION
**Pull request by Bibo-Joshi**
See #4125 for discussion

Since the tests for these things usually don't run well on Windows, I haven't run the unit tests yet - let's see.
Adding new unit tests will also likely be chellanging. If the test coverage doesn't drop, I might be content with leaving as is 😬

### CODE REVIEW

**Comment by Bibo-Joshi** [comment]
All right, tests seem to be stuck. will have to debug on linux

**Comment by Bibo-Joshi** [comment]
Test coverage seems to be fine and as expected adjusting the existing tests was already rather annoying, so tbh I'm not eager to try and add another unit test here.
I manually ran the customwebhook as well, which still works fine.

File: telegram/ext/_updater.py [hunk 1]
```diff
@@ -126,6 +126,7 @@ def __init__(
         self._httpd: Optional[WebhookServer] = None
         self.__lock = asyncio.Lock()
         self.__polling_task: Optional[asyncio.Task] = None
+        self.__polling_task_stop_event: asyncio.Event = asyncio.Event()
```
Comment (harshil21) [inline telegram/ext/_updater.py:129]: need the slot entry for this...

File: telegram/ext/_updater.py [hunk 2]
```diff
@@ -706,39 +709,58 @@ async def _network_loop_retry(
             description (:obj:`str`): Description text to use for logs and exception raised.
             interval (:obj:`float` | :obj:`int`): Interval to sleep between each call to
                 `action_cb`.
+            stop_event (:class:`asyncio.Event` | :obj:`None`): Event to wait on for stopping the
+                loop. Setting the event will make the loop exit even if action_cb is currently
```
Comment (harshil21) [inline telegram/ext/_updater.py:713]:
```suggestion
                loop. Setting the event will make the loop exit even if `action_cb` is currently
```

File: tests/ext/test_application.py [hunk 3]
```diff
@@ -1432,15 +1432,22 @@ async def callback(update, context):
...
 
         async def get_updates(*args, **kwargs):
             if exception_event.is_set():
                 raise exception
+
             # This makes sure that other coroutines have a chance of running as well
-            await asyncio.sleep(0)
+            if exception_testing_done.is_set() and app.updater.running:
+                # the longer sleep makes sure that we can exit also while get_updates is running
+                await asyncio.sleep(20)
```
Comment (harshil21) [inline tests/ext/test_application.py:1447]: does it really have to be 20 secs? Doesn't that mean the test suite is hung up by 20 secs?

File: tests/ext/test_updater.py [hunk 4]
```diff
@@ -611,8 +614,9 @@ async def get_updates(*args, **kwargs):
                 # Unfortunately we need to use the private attribute here to produce the problem
                 updater._running = False
                 second_update_event.set()
+                await asyncio.sleep(1)
+                # await third_update_event.wait()
```
Comment (harshil21) [inline tests/ext/test_updater.py:618]: don't need this?

File: telegram/ext/_updater.py [hunk 5]
```diff
@@ -706,39 +709,58 @@ async def _network_loop_retry(
...
+
+        async def do_action() -> bool:
+            if not stop_event:
+                return await action_cb()
+
+            action_cb_task = asyncio.create_task(action_cb())
+            stop_task = asyncio.create_task(stop_event.wait())
+            done, pending = await asyncio.wait(
+                (action_cb_task, stop_task), return_when=asyncio.FIRST_COMPLETED
+            )
```
Comment (harshil21) [inline telegram/ext/_updater.py:726]: clever

**Review by harshil21 — COMMENTED** [review]
code looks good. Didn't have a very close look at the tests.. they are kind of a black box because of so many `Event`'s. I assume you ran an example bot on both Windows and linux and checked if Ctrl+C works well?

File: tests/ext/test_application.py (hunk 3 above)
Comment (Bibo-Joshi) [reply tests/ext/test_application.py:1447]: no, what I'm testing here is that calling `Updater.stop` does indeed *not* wait for the 2 seconds to finish, i.e. that  `asyncio.wait` returns the `event.wait()` as done :)

**Comment by Bibo-Joshi** [comment]
> Didn't have a very close look at the tests.. they are kind of a black box because of so many `Event`'s.

They are certainly not very pretty :/

> I assume you ran an example bot on both Windows and linux and checked if Ctrl+C works well?

Yes, I ran echobot & the example from #4125.

**Review by harshil21 — APPROVED** [review]

**Pull request merged**